[feature] Migration#add_index

donaldong opened the pull request.

katyho commented on lib/redcord/migration/index.rb:
```diff
@@ -5,6 +5,41 @@
+
+      # Force refresh on the updated index attribute. Use multi to ensure
+      # consistency incase this migration failed in the middle
+      model.redis.multi
```
Should you start the multi command before the `hmget` so that you ensure the value doesn't change? 

What are the reasons for using Multi/Exec vs a lua script?

donaldong replied:
hmmm. yeah; it's possible that the index_val has changed when we call update_hash. I guess this has to be done via a lua script.

katyho commented on lib/redcord/migration/index.rb:
```diff
@@ -5,6 +5,41 @@
+      # Force refresh on the updated index attribute. Use multi to ensure
+      # consistency incase this migration failed in the middle
+      model.redis.multi
+      model.redis.update_hash(
+        model.model_key,
+        id,
+        {index_name => index_val.empty? ? 'redcord_temp_val' : ''},
```
I don't really understand this part. Isn't the code below sufficient? 

`update_hash` also resets the TTL for the record, and I'm not sure we want to do that if we're just indexing it and no client has actually touched it

donaldong replied:
the motivation here is to "force refresh" an attribute so the logic in update_hash can add things to the index structures.

katyho replied:
Why do you need to call `update_hash` twice? I think it might be better to duplicate or refactor the logic in update_hash so we don't reset the TTL on a record?

donaldong replied:
we have some logic to leave the index structure along unless the value has changed.

> so we don't reset the TTL on a record?

 indeed! we should have a separate script for adding index.

donaldong commented on lib/redcord/attribute.rb:
```diff
@@ -60,6 +60,7 @@ def ttl(duration)
 
     private
 
+    # TODO: remove me
```
oh no this is for the sadd_proc_on_redis_connection method haha

donaldong commented:
add a lua script

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/15f39dcb5332f41eacd8647eea7da291b71529f8..db58f87bcd3d4a883e72b0f5848b8a0cbfaae2d2).

donaldong commented on lib/redcord/server_scripts/add_index.erb.lua:
```diff
@@ -0,0 +1,46 @@
+if redis.call('sismember', model .. ':index_attrs', index_name) == 1 then
+  add_id_to_index_attr(model, index_name, index_val, id)
+elseif redis.call('sismember', model .. ':range_index_attrs', index_name) == 1 then
+  add_id_to_range_index_attr(model, index_name, index_val, id)
```
Good catch!

katyho reviewed (approved):
Overall LGTM! Few non-blocking comments

donaldong commented:
I will rebase this on #67
